### ISSUE TITLE
Parameter must be an array or an object that implements Countable

### ISSUE THREAD

**Issue by cocodrino** [guzzle/guzzle#2107]
Hi guzzle, I'm receiving this error on the line:

https://github.com/guzzle/guzzle/blob/7bc46be28e4b01b96d0dfda8c58169072e626558/src/Handler/CurlFactory.php#L66

I think that this solved changing this to
```php
if (is_array($this->handles)  && count($this->handles) >= $this->maxHandles) { //check if it's array first
            curl_close($resource);
        } else {
            // Remove all callback functions as they can hold onto references
            // and are not cleaned up by curl_reset. Using curl_setopt_array
            // does not work for some reason, so removing each one
            // individually.
            curl_setopt($resource, CURLOPT_HEADERFUNCTION, null);
            curl_setopt($resource, CURLOPT_READFUNCTION, null);
            curl_setopt($resource, CURLOPT_WRITEFUNCTION, null);
            curl_setopt($resource, CURLOPT_PROGRESSFUNCTION, null);
            curl_reset($resource);
            $this->handles[] = $resource;
        }

```

is it ok??? thank you

**Comment by cocodrino** [comment]
this is the message:

```php




Whoops, looks like something went wrong.
1/1 ErrorException in CurlFactory.php line 67: count(): Parameter must be an array or an object that implements Countable

    in CurlFactory.php line 67
    at HandleExceptions->handleError('2', 'count(): Parameter must be an array or an object that implements Countable', 'C:\xampp\htdocs\marketing\vendor\guzzlehttp\guzzle\src\Handler\CurlFactory.php', '67', array('easy' => object(EasyHandle), 'resource' => resource))
    at count(null) in CurlFactory.php line 67
    at CurlFactory->release(object(EasyHandle)) in CurlFactory.php line 107
    at CurlFactory::finish(object(CurlHandler), object(EasyHandle), object(CurlFactory)) in CurlHandler.php line 43
    at CurlHandler->__invoke(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Proxy.php line 28
    at Proxy::GuzzleHttp\Handler\{closure}(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Proxy.php line 51
    at Proxy::GuzzleHttp\Handler\{closure}(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in PrepareBodyMiddleware.php line 72
    at PrepareBodyMiddleware->__invoke(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Middleware.php line 30
    at Middleware::GuzzleHttp\{closure}(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in RedirectMiddleware.php line 68
    at RedirectMiddleware->__invoke(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Middleware.php line 59
    at Middleware::GuzzleHttp\{closure}(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in HandlerStack.php line 67
    at HandlerStack->__invoke(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Client.php line 275
    at Client->transfer(object(Request), array('synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false)) in Client.php line 123
    at Client->requestAsync('post', object(Uri), array('form_params' => array('secret' => '6LfyISoTAAAAAC0hJ916unwi0m_B0p7fAvCRK4Kp', 'remoteip' => '::1', 'response' => '03AEMEkEnEBWpwbTmBwYn0utViyKNfl48eRMPyBKtM6XE1dnzChxMZkOb64MrpZcgYTGJdINZ4ujx5ztPN3XQdRL0jlMewR8Xmd_p9-5InxfCyCICuo3LyBeZkrGUqyJGQzGUM4TPyRAKOmEC_D0-czuldWz0yM16PItsXEIv_2jaCi8cVoHoYKHtiLQSxiig3tWQkXYGLHxCKjdzcG-OolyxPmWKO3PuJNkxG9UrZ0Ht5ns08bCpzPi1rlDYsRoHnPMLL3rNnDmRto9Vw0pAHnei57jchwGL5QA'), 'synchronous' => true, 'handler' => object(HandlerStack), 'allow_redirects' => array('max' => '5', 'protocols' => array('http', 'https'), 'strict' => false, 'referer' => false, 'track_redirects' => false), 'http_errors' => true, 'decode_content' => true, 'verify' => true, 'cookies' => false, '_conditional' => array('User-Agent' => 'GuzzleHttp/6.2.1 curl/7.59.0 PHP/7.2.5'))) in Client.php line 129
    at Client->request('post', 'https://www.google.com/recaptcha/api/siteverify', array('form_params' => array('secret' => '6LfyISoTAAAAAC0hJ916unwi0m_B0p7fAvCRK4Kp', 'remoteip' => '::1', 'response' => '03AEMEkEnEBWpwbTmBwYn0utViyKNfl48eRMPyBKtM6XE1dnzChxMZkOb64MrpZcgYTGJdINZ4ujx5ztPN3XQdRL0jlMewR8Xmd_p9-5InxfCyCICuo3LyBeZkrGUqyJGQzGUM4TPyRAKOmEC_D0-czuldWz0yM16PItsXEIv_2jaCi8cVoHoYKHtiLQSxiig3tWQkXYGLHxCKjdzcG-OolyxPmWKO3PuJNkxG9UrZ0Ht5ns08bCpzPi1rlDYsRoHnPMLL3rNnDmRto9Vw0pAHnei57jchwGL5QA'), 'synchronous' => true)) in Client.php line 87
    at Client->__call('post', array('https://www.google.com/recaptcha/api/siteverify', array('form_params' => array('secret' => '6LfyISoTAAAAAC0hJ916unwi0m_B0p7fAvCRK4Kp', 'remoteip' => '::1', 'response' => '03AEMEkEnEBWpwbTmBwYn0utViyKNfl48eRMPyBKtM6XE1dnzChxMZkOb64MrpZcgYTGJdINZ4ujx5ztPN3XQdRL0jlMewR8Xmd_p9-5InxfCyCICuo3LyBeZkrGUqyJGQzGUM4TPyRAKOmEC_D0-czuldWz0yM16PItsXEIv_2jaCi8cVoHoYKHtiLQSxiig3tWQkXYGLHxCKjdzcG-OolyxPmWKO3PuJNkxG9UrZ0Ht5ns08bCpzPi1rlDYsRoHnPMLL3rNnDmRto9Vw0pAHnei57jchwGL5QA')))) in Tool.php line 791
    at Tool::checkReCaptcha(object(Request)) in AuthController.php line 101
    at AuthController->validateLogin(object(Request)) in AuthenticatesUsers.php line 59
    at AuthController->login(object(Request))
    at call_user_func_array(array(object(AuthController), 'login'), array(object(Request))) in Controller.php line 80
    at Controller->callAction('login', array(object(Request))) in ControllerDispatcher.php line 146
    at ControllerDispatcher->call(object(AuthController), object(Route), 'login') in ControllerDispatcher.php line 94
    at ControllerDispatcher->Illuminate\Routing\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 52
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in RedirectIfAuthenticated.php line 25
    at RedirectIfAuthenticated->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(RedirectIfAuthenticated), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 102
    at Pipeline->then(object(Closure)) in ControllerDispatcher.php line 96
    at ControllerDispatcher->callWithinStack(object(AuthController), object(Route), object(Request), 'login') in ControllerDispatcher.php line 54
    at ControllerDispatcher->dispatch(object(Route), object(Request), 'Acelle\Http\Controllers\Auth\AuthController', 'login') in Route.php line 174
    at Route->runController(object(Request)) in Route.php line 140
    at Route->run(object(Request)) in Router.php line 724
    at Router->Illuminate\Routing\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 52
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in NotLoggedIn.php line 37
    at NotLoggedIn->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(NotLoggedIn), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in VerifyCsrfToken.php line 64
    at VerifyCsrfToken->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(VerifyCsrfToken), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in ShareErrorsFromSession.php line 49
    at ShareErrorsFromSession->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(ShareErrorsFromSession), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in StartSession.php line 64
    at StartSession->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(StartSession), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in AddQueuedCookiesToResponse.php line 37
    at AddQueuedCookiesToResponse->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(AddQueuedCookiesToResponse), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in EncryptCookies.php line 59
    at EncryptCookies->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(EncryptCookies), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 102
    at Pipeline->then(object(Closure)) in Router.php line 726
    at Router->runRouteWithinStack(object(Route), object(Request)) in Router.php line 699
    at Router->dispatchToRoute(object(Request)) in Router.php line 675
    at Router->dispatch(object(Request)) in Kernel.php line 246
    at Kernel->Illuminate\Foundation\Http\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 52
    at Pipeline->Illuminate\Routing\{closure}(object(Request)) in CheckForMaintenanceMode.php line 44
    at CheckForMaintenanceMode->handle(object(Request), object(Closure))
    at call_user_func_array(array(object(CheckForMaintenanceMode), 'handle'), array(object(Request), object(Closure))) in Pipeline.php line 136
    at Pipeline->Illuminate\Pipeline\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 32
    at Pipeline->Illuminate\Routing\{closure}(object(Request))
    at call_user_func(object(Closure), object(Request)) in Pipeline.php line 102
    at Pipeline->then(object(Closure)) in Kernel.php line 132
    at Kernel->sendRequestThroughRouter(object(Request)) in Kernel.php line 99
    at Kernel->handle(object(Request)) in main.php line 55
    at require('C:\xampp\htdocs\marketing\public\main.php') in index.php line 61


```

I know it's related to php 7 ...

**Comment by malohtie** [comment]
try update guzzle i think it fixed

**Comment by sagikazarmark** [comment]
This isn't the first time this issue is reported, but I simply cannot find any codepath that would render the handles property null. I can see two options:

- There is a bug in PHP's `array_pop` function. Since it accesses the array by reference, I can imagine something going on there.
- Something accesses the factory instance through reflection and messes things up.

Can you reproduce the issue or it just happens from time to time?

**Comment by cocodrino** [comment]
hi..this happens always with my version of php 7...I just edited the code as show above and it works, maybe I can test it more if you want...

**Comment by Zemistr** [comment]
Maybe because of this? :thinking: 

http://php.net/manual/en/control-structures.foreach.php

**Note:**

In PHP 5, when foreach first starts executing, the internal array pointer is automatically reset to the first element of the array. This means that you do not need to call reset() before a foreach loop.

As foreach relies on the internal array pointer in PHP 5, changing it within the loop may lead to unexpected behavior.

In PHP 7, foreach does not use the internal array pointer.

**Comment by Zemistr** [comment]
Looks like it is already fixed. :)
https://github.com/guzzle/guzzle/blob/master/src/Handler/CurlFactory.php#L18

**Comment by gmponos** [comment]
Hi.. lately we worked on guzzle to add static analysis tools. Since we have declared that the `hanldes` property is an array we are hoping that static analysis will help us detect if there is any chance somewhere in the code this could be set us null...

This is an old issue and the version of guzzle is not mentioned here. I would be grateful if someone has seen this using the latest version could report back.
